### PR TITLE
Pcap delay

### PR DESCRIPTION
**Pull request by duk3luk3**
Because it's awesome.

### CODE REVIEW

File: lua/include/packet.lua [hunk 1]
```diff
@@ -666,6 +666,12 @@ end
 --- Raw packet type
 pkt.getRawPacket = packetCreate()
 
+--! Setter for raw packets
+--! @param data: raw packet data
+function pkt:setRawPacket(data)
+	ffi.copy(self.pkt.data, data)
+end
```
Comment (emmericp) [inline lua/include/packet.lua]: maybe set the packet size here as well?

File: lua/include/pcap.lua [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+end
+
+--! Writes a pcap record header.
+--! @param file: the file to write to
+--! @param buf: the packet buffer
+--! @param starttime: the time offset for timestamping
+function writeRecordHeader(file, buf, starttime)
+	--pcap record header
+	local pcapRecord = ffi.new(pcaprec_hdr_s)
+	local have_ts_in_packet = false -- TODO: buf:getTimestamp only works on NICs that support timestamping all packets and returns garbage otherwise (Paul1 pls fix)
```
Comment (emmericp) [inline lua/include/pcap.lua]: camelCase please

File: lua/include/pcap.lua [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	local ts = buf:getTimestamp()
+	if ts and ts ~= 0 and have_ts_in_packet then
+		ts = ts / 10^9 -- convert from ns to s
+		local ts_s = math.floor(ts)
+		local ts_us = ts % 1 * 1000
```
Comment (emmericp) [inline lua/include/pcap.lua]: camelCase

File: lua/include/pcap.lua [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	file:write(ffi.string(pcapRecord, ffi.sizeof(pcapRecord)))
+end
+
+--! Generate an iterator for pcap records.
+--! @param file: the pcap file
+--! @param rate: the tx link rate in Mbit per second if packets should have proper delays
+--! @return: iterator for the pcap records
+function readPcapRecords(file, rate)  
+	local pcapFile = readAs(file, pcap_hdr_s)
+	local is_ns_resolution = false
```
Comment (emmericp) [inline lua/include/pcap.lua]: camelCase

File: lua/include/pcap.lua [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	if pcapFile.magic_number == 0xA1B2C34D then
+		is_ns_resolution = true
```
Comment (emmericp) [inline lua/include/pcap.lua]: .

File: lua/include/pcap.lua [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+pcapReader = {}
+
+--! Generates a new pcapReader.
+--! @param filename: filename to open and read from
+--! @param rate: The rate of the link, if the packets are supposed to be replayed
+function pcapReader:newPcapReader(filename, rate)
+	rate = rate or 10000
+	local file = io.open(filename, "r")
+	 --TODO validy checks with more meaningfull errors in an extra function
```
Comment (emmericp) [inline lua/include/pcap.lua]: *meaningful

File: lua/include/pcap.lua [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+end
+
+pcapWriter = {}
+
+--! Generates a new pcapWriter.
+--! @param filename: filename to open and write to
+function pcapWriter:newPcapWriter(filename)
+	local file = io.open(filename, "w")
+	local time = monotonic_time() -- for offsetting timestamps
+	print("Offset timestamp: ",time)
```
Comment (emmericp) [inline lua/include/pcap.lua]: remove

File: lua/include/utils.lua [hunk 8]
```diff
@@ -134,20 +145,52 @@ ntoh = hton
...
-		return tonumber(tv.tv_sec) + tonumber(tv.tv_usec) / 10^6
+		sec, usec = gettimeofday()
+		return sec + usec / 10^6
 	end
+
+	local ts = ffi.new("struct timespec")
+
+--- Return the current monotonic clock time
+--- @return The time in seconds (as a double)
+	function monotonic_time()
```
Comment (emmericp) [inline lua/include/utils.lua]: camelCase

Comment (emmericp) [inline lua/include/utils.lua]: camelCase

File: lua/include/utils.lua [hunk 9]
```diff
@@ -134,20 +145,52 @@ ntoh = hton
...
+		ffi.C.clock_gettime(1, ts)
+		return ts.tv_sec + ts.tv_nsec / 10^9
+	end
+end
+
+--- Subtract timeval values
+--- @param x the "later" timeval (Minuend)
+--- @param y the "earlier" time (Subtrahend)
+--- @return The result in microseconds
+function timeval_uspan(x, y)
```
Comment (emmericp) [inline lua/include/utils.lua]: camelCase

File: lua/include/packet.lua (hunk 1 above)
Comment (duk3luk3) [inline lua/include/packet.lua]: I already had it there once but then I segfaulted myself.

File: lua/include/pcap.lua (hunk 2 above)
Comment (duk3luk3) [inline lua/include/pcap.lua]: okay.jpg

**Pull request merged**